src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -2604,5 +2608,10 @@ public override IEntityPersister GetEntityPersister(string entityName, object ob
 				}
 			}
 		}
+
+		void RemoveChildSession(ISession session)
+		{
+			Interlocked.CompareExchange(ref _childSession, null, session);
```

Comment:
Session is not thread safe, so we should not care about thread safety here.

I know there is a case where `Dispose` can be called from another thread: when a transaction in which the session was enlisted ends. But this case is buggy and doomed anyway: the child-session is already awaiting disposal (`TransactionContext.ShouldCloseSessionOnDistributedTransactionCompleted` being `true`) and should already be considered unusable for any other purpose than ending that transaction. So having it still returnable by its parent session is a trouble.

Being thread safe here will not help avoid troubles. Either we should ensure the child session is immediately removed from its parent on child dispose/close call, even when its disposal is postponed due to an ongoing transaction (`TransactionContext` not null), or we should cease postponing disposal when there is an ongoing transaction, as Hibernate do with `JdbcContext`. In both cases, we would no more be exposed to the transaction completion occurring on a distinct thread trouble.